#4061 seems like break running vite in dev mode

### Describe the bug

i was gonna do some testing in resolve playground with vite running in dev mode. then got following error.

```
Error: Cannot find module 'vite/dist/client/client.mjs'
Require stack:
- /Users/taylorliu/Projects/dev-vite/vite/packages/vite/dist/node/constants.js
- /Users/taylorliu/Projects/dev-vite/vite/packages/vite/dist/node/utils.js
- /Users/taylorliu/Projects/dev-vite/vite/packages/vite/dist/node/plugins/reporter.js
- /Users/taylorliu/Projects/dev-vite/vite/packages/vite/dist/node/build.js
- /Users/taylorliu/Projects/dev-vite/vite/packages/vite/dist/node/config.js
- /Users/taylorliu/Projects/dev-vite/vite/packages/vite/dist/node/index.js
- /Users/taylorliu/Projects/dev-vite/vite/packages/vite/dist/node/cli.js
- /Users/taylorliu/Projects/dev-vite/vite/packages/vite/bin/vite.js
```
it looks like it's caused by #4061.

### Reproduction

you can test it in any playground using vite in dev mode.

### System Info

```shell
System:
    OS: macOS 10.15.2
    CPU: (4) x64 Intel(R) Core(TM) i5-7267U CPU @ 3.10GHz
    Memory: 433.47 MB / 8.00 GB
    Shell: 3.2.2 - /usr/local/bin/fish
  Binaries:
    Node: 14.15.4 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.10 - /usr/local/bin/npm
  Browsers:
    Chrome: 91.0.4472.114
    Firefox: 75.0
    Safari: 13.0.4
```


### Used Package Manager

yarn

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).

--- a/packages/vite/package.json
+++ b/packages/vite/package.json
@@ -31,9 +31,7 @@
   "homepage": "https://github.com/vitejs/vite/tree/main/#readme",
   "scripts": {
     "predev": "rimraf dist",
-    "dev": "run-p dev-client dev-node",
-    "dev-client": "tsc -w --incremental --p src/client",
-    "dev-node": "tsc -w --incremental --p src/node",
+    "dev": "rollup -c -w",
     "prebuild": "rimraf dist && yarn lint",
     "build": "run-s build-bundle build-types",
     "build-bundle": "rollup -c",

--- a/packages/vite/rollup.config.js
+++ b/packages/vite/rollup.config.js
@@ -91,76 +91,92 @@ const sharedNodeOptions = {
 }
 
 /**
- * @type { import('rollup').RollupOptions }
+ *
+ * @param {boolean} isProduction
+ * @returns {import('rollup').RollupOptions}
  */
-const nodeConfig = {
-  ...sharedNodeOptions,
-  input: {
-    index: path.resolve(__dirname, 'src/node/index.ts'),
-    cli: path.resolve(__dirname, 'src/node/cli.ts')
-  },
-  external: [
-    'fsevents',
-    ...Object.keys(require('./package.json').dependencies)
-  ],
-  plugins: [
-    alias({
-      // packages with "module" field that doesn't play well with cjs bundles
-      entries: {
-        '@vue/compiler-dom': require.resolve(
-          '@vue/compiler-dom/dist/compiler-dom.cjs.js'
-        ),
-        'big.js': require.resolve('big.js/big.js')
-      }
-    }),
-    nodeResolve({ preferBuiltins: true }),
-    typescript({
-      target: 'es2019',
-      include: ['src/**/*.ts'],
-      esModuleInterop: true
-    }),
-    // Some deps have try...catch require of optional deps, but rollup will
-    // generate code that force require them upfront for side effects.
-    // Shim them with eval() so rollup can skip these calls.
-    shimDepsPlugin({
-      'plugins/terser.ts': {
-        src: `require.resolve('terser'`,
-        replacement: `require.resolve('vite/dist/node/terser'`
-      },
-      // chokidar -> fsevents
-      'fsevents-handler.js': {
-        src: `require('fsevents')`,
-        replacement: `eval('require')('fsevents')`
-      },
-      // cac re-assigns module.exports even in its mjs dist
-      'cac/dist/index.mjs': {
-        src: `if (typeof module !== "undefined") {`,
-        replacement: `if (false) {`
-      },
-      // postcss-import -> sugarss
-      'process-content.js': {
-        src: 'require("sugarss")',
-        replacement: `eval('require')('sugarss')`
-      },
-      'import-from/index.js': {
-        pattern: /require\(resolveFrom/g,
-        replacement: `eval('require')(resolveFrom`
-      },
-      'lilconfig/dist/index.js': {
-        pattern: /: require,/g,
-        replacement: `: eval('require'),`
-      }
-    }),
-    // Optional peer deps of ws. Native deps that are mostly for performance.
-    // Since ws is not that perf critical for us, just ignore these deps.
-    ignoreDepPlugin({
-      bufferutil: 1,
-      'utf-8-validate': 1
-    }),
-    commonjs({ extensions: ['.js'] }),
-    json(),
-    licensePlugin()
-  ]
+const createNodeConfig = (isProduction) => {
+  /**
+   * @type { import('rollup').RollupOptions }
+   */
+  const nodeConfig = {
+    ...sharedNodeOptions,
+    input: {
+      index: path.resolve(__dirname, 'src/node/index.ts'),
+      cli: path.resolve(__dirname, 'src/node/cli.ts')
+    },
+    external: [
+      'fsevents',
+      ...Object.keys(require('./package.json').dependencies),
+      ...(isProduction
+        ? []
+        : Object.keys(require('./package.json').devDependencies))
+    ],
+    plugins: [
+      alias({
+        // packages with "module" field that doesn't play well with cjs bundles
+        entries: {
+          '@vue/compiler-dom': require.resolve(
+            '@vue/compiler-dom/dist/compiler-dom.cjs.js'
+          ),
+          'big.js': require.resolve('big.js/big.js')
+        }
+      }),
+      nodeResolve({ preferBuiltins: true }),
+      typescript({
+        target: 'es2019',
+        include: ['src/**/*.ts'],
+        esModuleInterop: true
+      }),
+      // Some deps have try...catch require of optional deps, but rollup will
+      // generate code that force require them upfront for side effects.
+      // Shim them with eval() so rollup can skip these calls.
+      isProduction &&
+        shimDepsPlugin({
+          'plugins/terser.ts': {
+            src: `require.resolve('terser'`,
+            replacement: `require.resolve('vite/dist/node/terser'`
+          },
+          // chokidar -> fsevents
+          'fsevents-handler.js': {
+            src: `require('fsevents')`,
+            replacement: `eval('require')('fsevents')`
+          },
+          // cac re-assigns module.exports even in its mjs dist
+          'cac/dist/index.mjs': {
+            src: `if (typeof module !== "undefined") {`,
+            replacement: `if (false) {`
+          },
+          // postcss-import -> sugarss
+          'process-content.js': {
+            src: 'require("sugarss")',
+            replacement: `eval('require')('sugarss')`
+          },
+          'import-from/index.js': {
+            pattern: /require\(resolveFrom/g,
+            replacement: `eval('require')(resolveFrom`
+          },
+          'lilconfig/dist/index.js': {
+            pattern: /: require,/g,
+            replacement: `: eval('require'),`
+          }
+        }),
+      // Optional peer deps of ws. Native deps that are mostly for performance.
+      // Since ws is not that perf critical for us, just ignore these deps.
+      ignoreDepPlugin({
+        bufferutil: 1,
+        'utf-8-validate': 1
+      }),
+      commonjs({
+        extensions: ['.js'],
+        ignore: ['bufferutil', 'utf-8-validate']
+      }),
+      json(),
+      isProduction && licensePlugin()
+    ]
+  }
+
+  return nodeConfig
 }
 
 /**
@@ -360,4 +376,14 @@ function licensePlugin() {
   })
 }
 
-export default [envConfig, clientConfig, nodeConfig, terserConfig]
+export default (commandLineArgs) => {
+  const isDev = commandLineArgs.watch
+  const isProduction = !isDev
+
+  return [
+    envConfig,
+    clientConfig,
+    createNodeConfig(isProduction),
+    ...(isProduction ? [terserConfig] : [])
+  ]
+}

